File: 234.palindrome-linked-list.cpp
/*
 * @lc app=leetcode id=234 lang=cpp
 *
 * [234] Palindrome Linked List
 *
 * https://leetcode.com/problems/palindrome-linked-list/description/
 *
 * algorithms
 * Easy (38.05%)
 * Likes:    2819
 * Dislikes: 340
 * Total Accepted:    394.6K
 * Total Submissions: 1M
 * Testcase Example:  '[1,2]'
 *
 * Given a singly linked list, determine if it is a palindrome.
 * 
 * Example 1:
 * 
 * 
 * Input: 1->2
 * Output: false
 * 
 * Example 2:
 * 
 * 
 * Input: 1->2->2->1
 * Output: true
 * 
 * Follow up:
 * Could you do it in O(n) time and O(1) space?
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        if (!head || !head->next)
            return true;
        ListNode *slow = head, *fast = head, *pre = NULL, *nxt;
        while (fast != NULL && fast->next != NULL)
        {
            fast = fast->next->next;
            nxt = slow->next;
            slow->next = pre;
            pre = slow;
            slow = nxt;
        }
        if (fast)
            slow = slow->next;
        while (slow)
        {
            if (slow->val != pre->val)
                return false;
            slow = slow->next;
            pre = pre->next;
        }
        return true;
    }
};
// @lc code=end

File: 234.palindrome-linked-list/234.palindrome-linked-list.java
/*
 * @lc app=leetcode id=234 lang=java
 *
 * [234] Palindrome Linked List
 *
 * https://leetcode.com/problems/palindrome-linked-list/description/
 *
 * algorithms
 * Easy (44.28%)
 * Likes:    6569
 * Dislikes: 477
 * Total Accepted:    765.5K
 * Total Submissions: 1.7M
 * Testcase Example:  '[1,2,2,1]'
 *
 * Given the head of a singly linked list, return true if it is a
 * palindrome.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: head = [1,2,2,1]
 * Output: true
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: head = [1,2]
 * Output: false
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The number of nodes in the list is in the range [1, 10^5].
 * 0 <= Node.val <= 9
 * 
 * 
 * 
 * Follow up: Could you do it in O(n) time and O(1) space?
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null) {
            return true;
        }

        ListNode fast = head;
        ListNode slow = head;

        while (fast != null) {
            if (fast.next == null) {
                break;
            }
            fast = fast.next;

            slow = slow.next;

            if (fast.next == null) {
                break;
            }
            fast = fast.next;
        }

        ListNode prev = slow;
        slow = slow.next;
        while (slow != null) {
            ListNode t = slow.next;
            slow.next = prev;
            prev = slow;
            slow = t;
        }

        while (head != prev && head.next != prev) {
            if (head.val != prev.val) {
                return false;
            }

            head = head.next;
            prev = prev.next;
        }

        return head.val == prev.val;
    }
}
// @lc code=end

